### PR TITLE
errors: add a user-level assertion-failed exit code

### PR DESCRIPTION
**Pull request by Stebalien**
Previously, we only had:

1. SysErrIllegalInstruction for crashes/panics.
2. ErrUnspecified for any form of "unspecified" error from an actor.

However, deliberate assertions/panics fell somewhere between these two:

1. We shouldn't treat them as "illegal instructions" as that makes it difficult to distinguish between "true crashes" and "clean panics".
2. We shouldn't treat them as unspecified, as we _know_ that there's a bug where "unspecified" means "eh, I don't care enough to provide a good error."

So I've added a new `ErrAssertionFailed` (to mirror the system's `SysErrAssertionFailed`). Actors should return _this_ error when they:

1. Panic. I.e., actors should install a panic handler that aborts with this exit code.
2. Fail an assertion. In most languages, catching panics will cover this as well.

Etc.

### CODE REVIEW

**Review by anorth — APPROVED** [review]
Makes sense to me

File: 07-errors.md [hunk 1]
```diff
@@ -78,15 +80,17 @@ Standard non-zero exit codes are codes returned by actors to indicate common fai
 | 21    | `ErrSerialization`     | actor failed to serialize or deserialize some state |
 | 22    | `ErrUnhandledMessage`  | actor cannot handle this message                    |
 | 23    | `ErrUnspecified`       | the actor failed with an unspecified error          |
-| 24-31 | reserved               |                                                     |
+| 24    | `ErrAssertionFailed`   | the actor failed some internal assertion            |
+| 25-31 | reserved               |                                                     |
 
 Changes from pre-FVM Filecoin:
 
 1. Added `ExitCode::ErrUnhnadledMessage` (replacing `ExitCode::SysErrInvalidMethod`).
```
Comment (raulk) [inline 07-errors.md:88]: Noticed a typo:

```suggestion
1. Added `ExitCode::ErrUnhadledMessage` (replacing `ExitCode::SysErrInvalidMethod`).
```

File: 07-errors.md [hunk 2]
```diff
@@ -78,15 +80,17 @@ Standard non-zero exit codes are codes returned by actors to indicate common fai
...
-2. Added `ExitCode::ErrUnspecified`. This generally means "there's a bug and I have no idea what
-   went wrong". The alternative would be for the actor to _panic_ (leading to an
-   `ExitCode::ErrIllegalInstruction` syscall error), but it may be useful to distinguish between
-   "crashing" and "something unknown went wrong".
+2. Added `ExitCode::ErrUnspecified`. This indicates that something unspecified went wrong, but we
+   don't know what the cause is. This is a sane default exit code when accurately determining the
+   underlying error is expensive, unreliable, complex, etc.
+3. Added `ExitCode::ErrAssertionFailed`. This means that the actor violated some internal runtime
+   invariant. _Clean_ panics should exit with this code.
```
Comment (raulk) [inline 07-errors.md:93]: What is `runtime` here? I don't think this is correct. There are either system-originated exit codes, or user-originated exit codes. This is the latter. Yeah, the builtin-actors `Runtime` emits this, but the concept of a runtime is not standard and it's an implementation choice, so it shouldn't leak here.

File: 07-errors.md [hunk 3]
```diff
@@ -44,14 +44,16 @@ Notes:
    `ExitCode::SysErrSenderInvalid` exit code.
 2. If an actor calls `vm::abort` with a system exit code, that exit code will be replaced with
    `ExitCode::SysErrIllegalExitCode`.
-3. If an actor panics/crashes/etc., the exit code `ExitCode::SysErrIllegalInstruction` will be
-   returned.
+3. If an actor crashes/etc., the exit code `ExitCode::SysErrIllegalInstruction` will be
+   returned. _Clean_ panics (i.e., rust, go, etc. throws an exception) should cleanly abort with
+   `ExitCode::ErrAssertionFailed`. `ExitCode::SysErrIllegalInstruction` means something went
+   _really_ wrong.
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply 07-errors.md:50]: I've just linked to the wasm docs.

File: 07-errors.md (hunk 2 above)
Comment (Stebalien) [reply 07-errors.md:93]: I said "runtime" to distinguish between some _state_ invariant. But I agree that it's really confusing.

**Pull request merged**